### PR TITLE
CORS-2902: capi/aws: add ext-LB as CAPA secondary LB

### PR DESCRIPTION
**Pull request by r4f4**
This PR makes the internal LB be the primary CAPA load balancer and then adds the external LB as a secondary one when using public endpoints.

### CODE REVIEW

File: pkg/asset/manifests/aws/cluster.go [hunk 1]
```diff
@@ -151,19 +152,53 @@ func GenerateClusterAssets(installConfig *installconfig.InstallConfig, clusterID
...
+						ToPort:      22623,
+						CidrBlocks:  []string{mainCIDR.String()},
+					},
+					{
+						Description: "Kube API traffic",
+						Protocol:    capa.SecurityGroupProtocolTCP,
+						FromPort:    6443,
+						ToPort:      6443,
+						CidrBlocks:  []string{"0.0.0.0/0"},
+					},
```
Comment (mtulio) [inline pkg/asset/manifests/aws/cluster.go:178]: @r4f4 based in our conversation on Slack, do we need that allow-all by default here, or only when the publish is External?

File: pkg/asset/manifests/aws/cluster.go [hunk 2]
```diff
@@ -151,19 +152,53 @@ func GenerateClusterAssets(installConfig *installconfig.InstallConfig, clusterID
...
+			Scheme:                 &capa.ELBSchemeInternetFacing,
+			CrossZoneLoadBalancing: true,
+			IngressRules: []capa.IngressRule{
+				{
+					Description: "Kube API traffic",
+					Protocol:    capa.SecurityGroupProtocolTCP,
+					FromPort:    6443,
+					ToPort:      6443,
+					CidrBlocks:  []string{"0.0.0.0/0"},
+				},
```
Comment (mtulio) [inline pkg/asset/manifests/aws/cluster.go:197]: Follow up our Slack conversaion: is those rules (`IngressRules`) takes effect (is reconciled) in the SG used by `SecondaryControlPlaneLoadBalancer`?

**Review by mtulio — COMMENTED** [review]
Hey Rafael. Some open questions, otherwise LGTM

File: pkg/asset/manifests/aws/cluster.go [hunk 3]
```diff
@@ -151,19 +152,53 @@ func GenerateClusterAssets(installConfig *installconfig.InstallConfig, clusterID
...
 
+	if installConfig.Config.Publish == types.ExternalPublishingStrategy {
+		awsCluster.Spec.SecondaryControlPlaneLoadBalancer = &capa.AWSLoadBalancerSpec{
+			Name:                   ptr.To(clusterID.InfraID + "-ext"),
+			LoadBalancerType:       capa.LoadBalancerTypeNLB,
+			Scheme:                 &capa.ELBSchemeInternetFacing,
+			CrossZoneLoadBalancing: true,
+			IngressRules: []capa.IngressRule{
+				{
+					Description: "Kube API traffic",
```
Comment (mtulio) [inline pkg/asset/manifests/aws/cluster.go:192]: Maybe?

```suggestion
					Description: "Kubernetes API Server traffic for public access",
```

File: pkg/asset/manifests/aws/cluster.go [hunk 4]
```diff
@@ -151,19 +152,53 @@ func GenerateClusterAssets(installConfig *installconfig.InstallConfig, clusterID
...
+				IngressRules: []capa.IngressRule{
+					{
+						Description: "MCS traffic from cluster",
+						Protocol:    capa.SecurityGroupProtocolTCP,
+						FromPort:    22623,
+						ToPort:      22623,
+						CidrBlocks:  []string{mainCIDR.String()},
+					},
+					{
+						Description: "Kube API traffic",
```
Comment (mtulio) [inline pkg/asset/manifests/aws/cluster.go:173]:
```suggestion
						Description: "Kubernetes API Server traffic for internal access",
```

File: pkg/asset/manifests/aws/cluster.go [hunk 5]
```diff
@@ -151,19 +152,53 @@ func GenerateClusterAssets(installConfig *installconfig.InstallConfig, clusterID
...
+						Protocol:    capa.SecurityGroupProtocolTCP,
+						FromPort:    6443,
+						ToPort:      6443,
+						CidrBlocks:  []string{"0.0.0.0/0"},
```
Comment (mtulio) [inline pkg/asset/manifests/aws/cluster.go:177]:
```suggestion
						CidrBlocks:  []string{mainCIDR.String()},
```

File: pkg/asset/manifests/aws/cluster.go [hunk 6]
```diff
@@ -151,19 +152,53 @@ func GenerateClusterAssets(installConfig *installconfig.InstallConfig, clusterID
...
+				Scheme:           &capa.ELBSchemeInternal,
 				AdditionalListeners: []capa.AdditionalListenerSpec{
 					{
 						Port:     22623,
 						Protocol: capa.ELBProtocolTCP,
 					},
 				},
+				IngressRules: []capa.IngressRule{
+					{
+						Description: "MCS traffic from cluster",
```
Comment (mtulio) [inline pkg/asset/manifests/aws/cluster.go:166]:
```suggestion
						Description: "Machine Config Server traffic from internal cluster",
```

**Review by mtulio — COMMENTED** [review]
Suggestions in the rule names.

File: pkg/asset/manifests/aws/cluster.go (hunk 5 above)
Comment (mtulio) [reply pkg/asset/manifests/aws/cluster.go:177]: ^ based in the Slack thread :)

File: pkg/asset/manifests/aws/cluster.go (hunk 1 above)
Comment (r4f4) [reply pkg/asset/manifests/aws/cluster.go:178]: Good point. Will fix.

Comment (r4f4) [reply pkg/asset/manifests/aws/cluster.go]: Fixed.

File: pkg/asset/manifests/aws/cluster.go (hunk 3 above)
Comment (r4f4) [reply pkg/asset/manifests/aws/cluster.go]: Fixed

File: pkg/asset/manifests/aws/cluster.go (hunk 4 above)
Comment (r4f4) [reply pkg/asset/manifests/aws/cluster.go]: Fixed.

File: pkg/asset/manifests/aws/cluster.go (hunk 6 above)
Comment (r4f4) [reply pkg/asset/manifests/aws/cluster.go]: Fixed.

**Comment by r4f4** [comment]
Update: addressed review comments.

File: pkg/asset/manifests/aws/cluster.go [hunk 7]
```diff
@@ -151,19 +152,53 @@ func GenerateClusterAssets(installConfig *installconfig.InstallConfig, clusterID
...
+						FromPort:    6443,
+						ToPort:      6443,
+						CidrBlocks:  []string{mainCIDR.String()},
+					},
+				},
 			},
 		},
 	}
 
+	if installConfig.Config.Publish == types.ExternalPublishingStrategy {
```
Comment (mtulio) [inline pkg/asset/manifests/aws/cluster.go:184]: @r4f4 here is the changes I needed to apply considering the CAPA limitations* we are discussing on Slack.

```suggestion
				IngressRules: []capa.IngressRule{
					{
						Description: "Machine Config Server internal traffic from cluster",
						Protocol:    capa.SecurityGroupProtocolTCP,
						FromPort:    22623,
						ToPort:      22623,
						CidrBlocks:  []string{mainCIDR.String()},
					},
				},
			},
		},
	}

	if installConfig.Config.Publish == types.ExternalPublishingStrategy {
	    awsCluster.Spec.ControlPlaneLoadBalancer.IngressRules = append(awsCluster.Spec.ControlPlaneLoadBalancer.IngressRules, capa.IngressRule{
			Description: "Kubernetes API Server traffic for public access",
			Protocol:    capa.SecurityGroupProtocolTCP,
			FromPort:    6443,
			ToPort:      6443,
			CidrBlocks:  []string{"0.0.0.0/0"},
		})
```

*Limitations we found so far:
- When defining the `ControlPlaneLoadBalancer.IngressRules` for API w/ internal access, the CAPA will complain due duplicated rule. CAPA Automatically add the ingress rule <VPC CIDR> 6443/TCP
- When defining the `SecondaryControlPlaneLoadBalancer.IngressRules` the CAPA is ignoring

**Review by mtulio — COMMENTED** [review]
@r4f4 I added this rule to the deployment considering the limitations in when syncing the rules for the SG shared in both LBs

File: pkg/asset/manifests/aws/cluster.go [hunk 8]
```diff
@@ -151,19 +152,62 @@ func GenerateClusterAssets(installConfig *installconfig.InstallConfig, clusterID
...
+				ToPort:      6443,
+				CidrBlocks:  []string{"0.0.0.0/0"},
+			},
+		)
+
+		awsCluster.Spec.SecondaryControlPlaneLoadBalancer = &capa.AWSLoadBalancerSpec{
+			Name:                   ptr.To(clusterID.InfraID + "-ext"),
+			LoadBalancerType:       capa.LoadBalancerTypeNLB,
+			Scheme:                 &capa.ELBSchemeInternetFacing,
+			CrossZoneLoadBalancing: true,
```
Comment (patrickdillon) [inline pkg/asset/manifests/aws/cluster.go:198]: we need to enable cross zone load balancing on the primary as well, but we could do that in a follow up pr if preferred

File: pkg/asset/manifests/aws/cluster.go [hunk 9]
```diff
@@ -151,19 +152,62 @@ func GenerateClusterAssets(installConfig *installconfig.InstallConfig, clusterID
...
 			},
 		},
 	}
 
+	if installConfig.Config.Publish == types.ExternalPublishingStrategy {
+		// FIXME: CAPA bug. Remove when fixed upstream
+		// The primary and secondary load balancers in CAPA share the same
+		// security group. However, specifying an ingress rule only in the
+		// second LB does not seem to take effect, forcing us to add it to the
+		// primary LB instead.
```
Comment (vincepri) [inline pkg/asset/manifests/aws/cluster.go:182]: Is there a PR/issue that we can link to?
cc @nrb

File: pkg/asset/manifests/aws/cluster.go (hunk 8 above)
Comment (vincepri) [reply pkg/asset/manifests/aws/cluster.go:198]: should we enable it only if there is more than 1 subnet setup?

File: pkg/asset/manifests/aws/cluster.go (hunk 9 above)
Comment (r4f4) [reply pkg/asset/manifests/aws/cluster.go:182]: I'll create one in a moment.

Comment (nrb) [reply pkg/asset/manifests/aws/cluster.go:182]: Thanks for https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/4865 @r4f4. Assigned myself to the upstream bug.

Comment (r4f4) [reply pkg/asset/manifests/aws/cluster.go:182]: https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4866

Comment (nrb) [reply pkg/asset/manifests/aws/cluster.go:182]: Posted my reply before seeing your PR. Thanks, will review.

Comment (r4f4) [reply pkg/asset/manifests/aws/cluster.go:182]: Added a link to the upstream issue in the comments.

File: pkg/asset/manifests/aws/cluster.go (hunk 2 above)
Comment (r4f4) [reply pkg/asset/manifests/aws/cluster.go]: The answer is "no" and we created https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/4865

File: pkg/asset/manifests/aws/cluster.go (hunk 8 above)
Comment (r4f4) [reply pkg/asset/manifests/aws/cluster.go:199]: In terraform we always enable it. I'm going to enable in both and we can review that decision in a follow-up.

**Review by mtulio — COMMENTED** [review]
/lgtm

**Comment by r4f4** [comment]
`altinfra-e2e-aws-capi-ovn` is expected to fail because of an issue with feature gates in the nightly images.
/override ci/prow/altinfra-e2e-aws-capi-ovn

**Pull request merged**